### ISSUE TITLE
Scala 2.12

### ISSUE THREAD

**Issue by filinep** [storm-enroute/coroutines#44]
Can we please get a release for scala 2.12?

**Comment by maximiliancsuk** [comment]
I would also really appreciate it if there is a build for 2.12 :)

**Comment by axel22** [comment]
Until [this specialization bug](https://github.com/scala/bug/issues/10277), which prevents the specialization of primitive types and would cause boxing for coroutines that yield primitive values, is fixed in Scala 2.12, I am reluctant to release a 2.12 version.

If you really need it before then, I would suggest building/releasing a 2.12 version privately and using that.

**Comment by maximiliancsuk** [comment]
That makes a lot of sense! Thanks for clarifying!

**Comment by Sciss** [comment]
Hi there. What's the status of this project? Is it dead or alive? What happens for Scala 2.12 now?

**Comment by axel22** [comment]
Hi! The project is temporarily put on hold until it becomes more clear what shape will Scala Macros take after 2.12. Until that, I would suggest forking, disabling specialization in the build flags, setting Scala version to 2.12 and releasing privately.

While not strictly necessary, it would help if [this](https://github.com/scala/bug/issues/10277) specialization bug from 2.12 were fixed (it is performance-related).

**Comment by Sciss** [comment]
@axel22 thanks; so that's good news in a way, as it means the project can be compiled. I'll have a look then!

**Comment by maximiliancsuk** [comment]
@Sciss would you mind sharing once you get a 2.12 build working? That'd be awesome! :)

**Comment by axel22** [comment]
@Sciss Yes, it should be compiled (current build failures are Travis-related). But if it does not for any reason, please let me know, and I will fix it for you.

**Comment by Sciss** [comment]
I have managed to get things mostly to compile (in 2.11 now) without mecha and so on, but I'm running into trouble with `JBench`. That class doesn't seem to be part of stable scalameter-core 0.8.2. Where do I get this dependency from?

**Comment by axel22** [comment]
It should be in scalameter-0.8.2: https://mvnrepository.com/artifact/com.storm-enroute/scalameter_2.11/0.8.2

Could you perhaps see if setting the version of ScalaMeter to 0.8.2 works? If not, please let me know and I will fix this.

https://github.com/storm-enroute/coroutines/blob/master/dependencies.conf

**Comment by Sciss** [comment]
No, it didn't work, also the `api` package is not present; but I use 0.9-SNAPSHOT now, it works, and it's fine because it's only a test dependency. I publish my new sbt build soon.

**Comment by axel22** [comment]
Ok, thanks a lot!

**Comment by Sciss** [comment]
Here is my work so far: https://github.com/Sciss/coroutines/tree/work

I updated and changed the build (using `build.sbt` now). I had to remove one or two tests, otherwise `sbt +clean +test` seems to be fine with both 2.11 and 2.12. I also ran through various code cleanups in IntelliJ, like adding return types, removing procedure syntax and such.

**Comment by axel22** [comment]
Looks pretty good! It's unfortunate that the boxing tests no longer work on 2.12, but let's hope that the way specialization interacts with JDK8-style mixin composition gets fixed.

**Comment by maximiliancsuk** [comment]
@Sciss I successfully built your codebase for 2.12 on my machine! :) I needed to add a few missing libraries and had to change a couple compile flags though. I created a pull request.

I'd be awesome if this would flow back into the repo here eventually! :)

**Comment by axel22** [comment]
I can create a 2.12 branch, if you wish to push the changes back.

**Comment by Sciss** [comment]
A 2.12 branch would probably be the best; I changed quite a bit in the code style, following IntelliJ and scalac warnings, like procedure syntax, explicit return types, unused statements and such. There were also a few tiny bugs, like missing string interpolation. I had to replace `build/Build.scala` with a simpler `./build.sbt`, so you might want to fusion that with your build system. Other than that, I commented out one or two tests that failed in Scala 2.12. I could create a pull request against a new branch that you suggest, and then you could selectively merge back into master or so.

**Comment by axel22** [comment]
Branch created: https://github.com/storm-enroute/coroutines/tree/2.12.x

Feel free to push a PR to it! And thnx!
